Produce docker image warning

StefanoFioravanzo opened the pull request:
When clicking the compile/build/deploy button, check if the current notebook image is different from the NotebookServer's one. In that case, produce a warning dialog informing the user, asking whether to continue or not.

@tasos-ale Since `NotebookUtils.showYesNoDialog` is `async`, this makes the function `async`. Is it ok?

elikatsis commented on src/components/LeftPanelWidget.tsx:
```diff
@@ -538,10 +538,38 @@ export class KubeflowKaleLeftPanel extends React.Component<IProps, IState> {
+        'Proceed anyway?',
+      ];
+      proceed = await NotebookUtils.showYesNoDialog(
+        'New NotebookServer Image Detected',
```
```suggestion
        'Notebook server's image is different!',
```

elikatsis commented on src/components/LeftPanelWidget.tsx:
```diff
@@ -538,10 +538,38 @@ export class KubeflowKaleLeftPanel extends React.Component<IProps, IState> {
+        '',
+        '   Current notebook metadata image: ' +
+          this.state.metadata.docker_image,
+        '   New NotebookServer image:        ' +
```
```suggestion
        '   Notebook server's image:        ' +
```

elikatsis commented on src/components/LeftPanelWidget.tsx:
```diff
@@ -538,10 +538,38 @@ export class KubeflowKaleLeftPanel extends React.Component<IProps, IState> {
+        '   New NotebookServer image:        ' +
+          DefaultState.metadata.docker_image,
+        '',
+        "In order to use this NotebookServer's image as base image for the" +
```
```suggestion
        "In order to use this notebook server's image as base image for the" +
```

elikatsis commented on src/components/LeftPanelWidget.tsx:
```diff
@@ -538,10 +538,38 @@ export class KubeflowKaleLeftPanel extends React.Component<IProps, IState> {
+    ) {
+      const body = [
+        'This Notebook was last run with a different image than the one used' +
+          ' by the current Notebook Server',
```
(unformatted)
```suggestion
        'This notebook is configured to run with a different image than the' +
          ' one used by the current notebook server',
```

elikatsis commented on src/components/LeftPanelWidget.tsx:
```diff
@@ -538,10 +538,38 @@ export class KubeflowKaleLeftPanel extends React.Component<IProps, IState> {
+      );
+    }
+    if (proceed) {
+      if (!this.state.runDeployment) {
```
I think this check should be at the top like:

```ts
if (this.state.runDeployment) {
  return;
}
```
and then all the rest

StefanoFioravanzo commented:
@elikatsis I addressed your comments and changed a bit the formatting of the message. Also, I grouped together the body generation of `NotebookUtils.showMessage` and `NotebookUtils.showYesNoDialog` in `NotebookUtils.buildDialogBody`

elikatsis commented on src/utils/NotebookUtils.tsx:
```diff
@@ -2,10 +2,38 @@ import { Dialog, showDialog } from '@jupyterlab/apputils';
+   * @param msg A list of string with HTML formatting
+   * @returns a HTMLDivElement composed of a list of spans with formatted text
+   */
+  public static buildDialogBody(msg: string[]): ReactElement {
```
nit: Make this private?

elikatsis commented on style/index.css:
```diff
@@ -30,6 +30,7 @@
 }
 
 .kale-header-switch {
+  flex: 0 0 auto;
```
nit: Duplicate

elikatsis commented on src/components/LeftPanelWidget.tsx:
```diff
@@ -272,7 +272,12 @@ export class KubeflowKaleLeftPanel extends React.Component<IProps, IState> {
+        ...this.state.metadata,
+        docker_image: name || DefaultState.metadata.docker_image,
+      },
+    });
```
I suggest we drop that change and perform a replacement (if needed) once the button is pressed.
Leaving it as is would still produce inconsistent behavior when deleting the text of the field.
Since there's a placeholder, users will always know what happens when the field is empty.

Once the button is pressed we now deepcopy notebook's metadata.
If docker image [in that copy] is empty, replace it with `DefaultState.metadata.docker_image`.

StefanoFioravanzo commented:
@elikatsis Now the docker image is updated on deploy just when the current state has an empty value, just like we discussed. Moreover, thanks to @tasos-ale 's advice, now the placeholder value work correctly.

For the placeholder value: note that I had to set the `InputLabelProps` to true when the placeholder is present. Might be worth remembering for the future. https://github.com/kubeflow-kale/jupyterlab-kubeflow-kale/pull/42/commits/9ed47684b247c92a599fef663219d744828a0f45#diff-12f5ae63fcc38a513b1de1cde98c411cR141

Commit:
Multiline body in yes-no dialog

Commit:
Use SanitizedHTML to show formatted text in dialogs

Commit:
Add default docker image as placeholder

Commit:
Set docker image to default when empty

Commit:
Add warning badges and compile entry in deploy progress

Signed-off-by: Stefano Fioravanzo <stefano@arrikto.com>

Commit:
Drop duplicate line in css

Commit:
Show warning badge on compile action

Also, this commit removes all the small dialogs produced during compile,
upload and run actions, since the information is already displayed on
the bottom left.

Signed-off-by: Stefano Fioravanzo <stefano@arrikto.com>
